### PR TITLE
Fix serializers configuration priority

### PR DESCRIPTION
**Pull request by GuilhemN**
See https://github.com/FriendsOfSymfony/FOSRestBundle/pull/1611

### CODE REVIEW

File: Tests/DependencyInjection/Compiler/SerializerConfigurationPassTest.php [hunk 1]
```diff
@@ -62,24 +62,24 @@ public function testShouldConfigureJMSSerializer()
...
+    public function testJmsSerializerServiceSupersedesSerializerService()
     {
         $this->container->register('jms_serializer.serializer', 'JMS\Serializer\Serializer');
         $this->container->register('serializer', 'Symfony\Component\Serializer\Serializer');
 
         $compiler = new SerializerConfigurationPass();
         $compiler->process($this->container);
 
-        $this->assertSame('fos_rest.serializer.symfony', (string) $this->container->getAlias('fos_rest.serializer'));
+        $this->assertSame('fos_rest.serializer.jms', (string) $this->container->getAlias('fos_rest.serializer'));
```
Comment (GuilhemN) [inline Tests/DependencyInjection/Compiler/SerializerConfigurationPassTest.php]: @xabbuh this was wrong according to the docs but it is technically a bc break. Is it fine for you anyway?

Comment (lyrixx) [inline Tests/DependencyInjection/Compiler/SerializerConfigurationPassTest.php]: This is a huge BC break unfortunately. But acutally, I prefer this versio of the code... ;)

File: .styleci.yml [hunk 2]
```diff
@@ -1,3 +1,4 @@
 preset: symfony
 disabled:
   - braces
+  - php_unit_fqcn_annotation
```
Comment (phansys) [inline .styleci.yml]: This change doesn't seem related to the subject.

File: DependencyInjection/Configuration.php [hunk 3]
```diff
@@ -59,12 +59,14 @@ public function getConfigTreeBuilder()
                 ->arrayNode('access_denied_listener')
                     ->canBeEnabled()
                     ->beforeNormalization()
-                        ->ifArray()->then(function ($v) { if (!empty($v) && empty($v['formats'])) {
-     unset($v['enabled']);
-     $v = ['enabled' => true, 'formats' => $v];
- }
+                        ->ifArray()->then(function ($v) {
+                            if (!empty($v) && empty($v['formats'])) {
```
Comment (phansys) [inline DependencyInjection/Configuration.php]: Why not simply `$v && !$v['formats']`?

File: DependencyInjection/Configuration.php [hunk 4]
```diff
@@ -59,12 +59,14 @@ public function getConfigTreeBuilder()
...
+                                unset($v['enabled']);
```
Comment (phansys) [inline DependencyInjection/Configuration.php]: `$v['enabled']` is overridden at the next line, which is the reason for this `unset()` call?

File: DependencyInjection/Configuration.php [hunk 5]
```diff
@@ -59,12 +59,14 @@ public function getConfigTreeBuilder()
                 ->arrayNode('access_denied_listener')
                     ->canBeEnabled()
                     ->beforeNormalization()
-                        ->ifArray()->then(function ($v) { if (!empty($v) && empty($v['formats'])) {
-     unset($v['enabled']);
-     $v = ['enabled' => true, 'formats' => $v];
- }
+                        ->ifArray()->then(function ($v) {
```
Comment (phansys) [inline DependencyInjection/Configuration.php]: `$v` argument could be declared as `array`.

File: .styleci.yml (hunk 2 above)
Comment (GuilhemN) [inline .styleci.yml]: styleci failed without it

File: DependencyInjection/Configuration.php (hunk 5 above)
Comment (GuilhemN) [inline DependencyInjection/Configuration.php]: that's not related, just a cs fix

File: DependencyInjection/Configuration.php (hunk 3 above)
Comment (GuilhemN) [inline DependencyInjection/Configuration.php]: that's not related, please open another issue

**Comment by xabbuh** [comment]
@GuilhemN I would prefer if we could split this PR into two commits. One for the code style changes and one for the actual fix which would make it easier to use git blame in the future.

**Comment by GuilhemN** [comment]
@xabbuh sure, it's done.

File: DependencyInjection/Compiler/SerializerConfigurationPass.php [hunk 6]
```diff
@@ -41,31 +41,26 @@ public function process(ContainerBuilder $container)
...
         }
 
-        if ($container->has('jms_serializer.serializer')) {
-            $container->setAlias('fos_rest.serializer', 'fos_rest.serializer.jms');
+        // Remove the exception normalizer linked to the jms normalizer
+        $container->removeDefinition('fos_rest.serializer.exception_normalizer.jms');
+
+        // As there is no `jms_serializer.serializer` service, there is a `serializer` service
+        $class = $container->getParameterBag()->resolveValue(
+            $container->findDefinition('serializer')->getClass()
```
Comment (GuilhemN) [inline DependencyInjection/Compiler/SerializerConfigurationPass.php]: This is already managed [here](https://github.com/FriendsOfSymfony/FOSRestBundle/pull/1622/files/8d8f4452e049c0cf2e9f59b2ef8a99486447868a..717cab75b4a5567942cd024ca5e86468fd2b113a#diff-b4c2bb665099019e5209acdfe6371d2aR41).

File: DependencyInjection/Compiler/SerializerConfigurationPass.php [hunk 7]
```diff
@@ -41,31 +41,26 @@ public function process(ContainerBuilder $container)
...
+        );
+
+        if (is_subclass_of($class, 'Symfony\Component\Serializer\SerializerInterface')) {
+            $container->setAlias('fos_rest.serializer', 'fos_rest.serializer.symfony');
+        } elseif (is_subclass_of($class, 'FOS\RestBundle\Serializer\Serializer')) {
+            $container->setAlias('fos_rest.serializer', 'serializer');
```
Comment (xabbuh) [inline DependencyInjection/Compiler/SerializerConfigurationPass.php]: What about the case of `JMS\Serializer\SerializerInterface`?

Comment (GuilhemN) [inline DependencyInjection/Compiler/SerializerConfigurationPass.php]: It is already supported under `jms_serializer.serializer`. I don't think we need to support the case where the jms serializer is only configured under `serializer`. Wdyt?

Comment (GuilhemN) [inline DependencyInjection/Compiler/SerializerConfigurationPass.php]: I don't think it has ever been supported, what's supported is [removing the `serializer` alias](https://github.com/schmittjoh/JMSSerializerBundle/blob/master/DependencyInjection/Configuration.php#L45).

Comment (xabbuh) [inline DependencyInjection/Compiler/SerializerConfigurationPass.php]: 👍 then for me

File: DependencyInjection/Configuration.php (hunk 4 above)
Comment (GuilhemN) [inline DependencyInjection/Configuration.php]: Good question, it should probably be removed.

**Comment by GuilhemN** [comment]
Thanks for your reviews !
